halotools/empirical_models/abunmatch/engines/bin_free_cam_kernel.pyx

```diff
@@ -110,11 +110,35 @@ def setup_initial_indices(int iy, int nwin, int npts):
     return iy, init_iy_low, init_iy_high
 
 
+@cython.boundscheck(False)
+@cython.nonecheck(False)
+@cython.wraparound(False)
+cdef int _find_index(int[:] arr, int val):
+    for i in range(len(arr)):
+        if arr[i] == val:
+            return i
+    return -1
+
+
+@cython.boundscheck(False)
+@cython.nonecheck(False)
+@cython.wraparound(False)
+def get_value_at_rank(double[:] sorted_values, int rank1, int nwin, int add_subgrid_noise):
```

Comment:
Ok, quick implementation comment about this `get_value_at_rank` function.  This function is declared with `def` rather than `cdef`, which means `get_value_at_rank` returns a python object. If you run `cython -a bin_free_cam_kernel.pyx` and inspect the html output, you will see deep yellow on line 270. Click on it and see all the code cython generates to compile this line. To fix this, `get_value_at_rank` needs to be C-declared. See, e.g., the definition of `_bisect_left_kernel`. With this change, this means that `get_value_at_rank` will no longer be accessible by the outside world, but only accessible within cython, so you will not be permitted to add it to `__all__`. 

In general, it's good to get in the habit of always running `cython -a` whenever reviewing cython code. Because cython is a superset of python, the compiler will allow you to write code with avoidable python-layer interactions without complaining, so you just have to be vigilant and always check this manually.

Reply:
I `def`ed it so that it would be accessible to the code that does the left and right edges. 

But maybe a better way would be to have an internal `cdef int _get_value_at_rank` and then an external `def get_value_at_rank` that just wraps this?

Running `cython -a` is definitely a good idea

Reply:
`bin_free_cam_kernel.html` looks clean as of this commit. I also just did a simple timing test, comparing your branch to the current version of master, and it looks like there is a nearly negligible performance increase from this feature, so I think this PR is all set once it passes CI.

Reply:
Sweet, thanks for the comments!